slider/collection.py

```diff
@@ -0,0 +1,98 @@
+# TODO: not use replay private functions
+from .replay import _consume_int, _consume_string
+
+
+class CollectionDB:
+    """An osu! ``collection.db`` file.
+
+    Parameters
+    ----------
+    version : int
+        Version number (e.g. 20150203)
+    num_collections : int
+        Number of collections
+    collections : list[Collection]
+        List of :class:`~slider.collection.Collection` s
+    """
+    def __init__(self, version, num_collections, collections):
+        self.version = version
+        self.num_collections = num_collections
+        self.collections = collections
+
+    @classmethod
+    def from_path(cls, path):
+        """Read in a ``collection.db`` file from disk.
+
+        Parameters
+        ----------
+        path : str or pathlib.Path
+            The path to the file to read from.
+        """
+        with open(path, 'rb') as f:
+            return cls.from_file(f)
+
+    @classmethod
+    def from_file(cls, file):
+        """Read in a ``collection.db`` file from an open file object.
+
+        Parameters
+        ----------
+        file : file-like
+            The file object to read from.
+        """
+        return cls.parse(file.read())
+
+    @classmethod
+    def parse(cls, data):
+        """Parse from ``collection.db`` data.
+
+        Parameters
+        ----------
+        data : bytes
+            The data from a ``collection.db`` file.
+        """
+        buffer = bytearray(data)
+
+        version = _consume_int(buffer)
+        num_collections = _consume_int(buffer)
+        collections = []
+        for i in range(num_collections):
+            collections.append(Collection.parse(buffer))
+
+        return cls(version, num_collections, collections)
+
+
+class Collection:
+    """An osu! collection. One or more collections are present in a
+    `collection.db` file.
+
+    Parameters
+    ----------
+    name : str
+        The collection's name
+    num_beatmaps : int
+        How many beatmaps the collection contains
+    md5_hashes : list[str]
+        List of MD5 hashes of each beatmap
+    """
+    def __init__(self, name, num_beatmaps, md5_hashes):
+        self.name = name
+        self.num_beatmaps = num_beatmaps
```

Comment:
do we still need this int, or is it just `len(md5_hashes)`?

Reply:
It is a variable specified by the collection.db format so I leave it in for completeness. 
(If `num_beatmaps != len(md5_hashes)` then parsing will fail.)

Reply:
I think leaving this as an argument just allows the possibility that it becomes out of sync or is passed incorrectly. This seems like something similar to `slider.replay._consume_actions`, where the file format specifies a count, then `count` values. We only use the count when parsing, then return a list instead. A user who needs the `num_beatmaps` value could just do `len(collection.beatmaps)`, so it is still accessible.

Reply:
I can change it but I still think the API should match the `collection.db` format whenever possible